### PR TITLE
Inline autofill with new JS communication mechanism

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/0/1202370114120269/f

### Description
Switching away from using `window.postMessage` to injecting in configuration and available input types directly.

### Not included (to be done in future PRs)

- Handling the "update saved login" flow"
- Pretty UI
- Improvements to credential management screen
- More tests

### Steps to test this PR
FOR FIRST TESTS, ENSURE DEVICE AUTH IS **ENABLED**

#### Test no Autofill where you have not saved credentials
- [x]  Visit site with a login (e.g., https://trello.com/login)
- [x]  Click on email field; verify no autofill prompts 

#### Test Autofill offers to save when logging in
- [x]  Visit site with a login (e.g., https://trello.com/login) (tip: don't have to use valid login details on trello to save dialog)
- [x]  Enter login details and press "log in" button 
- [x] Verify dialog shown asking if you want to save login details 
- [x] Choose to save, then logout/refresh the page
- [x] Click on email field; verify that you do see the autofill prompt

#### Test Autofill offers to save when logging in
- [x]  Visit trello login page https://trello.com/login
- [x]  Enter login details and press "log in" button 
- [x] Verify dialog shown asking if you want to save login details 
- [x] Choose not to save, refresh and verify credentials not offered to autofill
- [x] Repeat the above steps, but this time choose to save, then logout/refresh the page
- [x] Click on email field; verify that you do see the autofill prompt

#### Test login credentials autofilled when chosen
- [x] Carrying on from last test where you'll have saved login credentials for trello...
- [x] Enter login details and press "log in" button 
- [x] Verify you see dialog with saved login shown
- [x] Choose to use the saved login
- [x] Verify you see the auth prompt. 
- [x] Let it fail
- [x] Verify autofill doesn't happen
- [x] Repeat the above (you'll have to refresh the page), this time let auth succeed
- [x] Verify autofill does happen

FOR THESE TESTS, ENSURE DEVICE AUTH IS **DISABLED**

#### Test login credentials autofilled when chosen

- [x] Repeat tests above, verifying that autofill prompts are never shown

### CODE REVIEW

**Review by karlenDimla — APPROVED** [review]
Aside from how the commits are looking here, the changes make sense. Found a few nits and improvements but nothing big.

I noticed the following though:
- When I add a duckduckgo email throught Email protection -> i already have a duck address -> even if I have entered the passphrase it doesn't move forward though the duckduckgo email is successfully added.

I don't think it is related to your changes but I only found this behavior in this PR and not in the nightly.

File: modules/autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/ui/credential/management/AutofillManagementActivity.kt [hunk 1]
```diff
@@ -117,6 +111,7 @@ class AutofillManagementActivity : DuckDuckGoActivity() {
     }
 
     private fun processState(state: AutofillSettingsViewModel.ViewState) {
+
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply modules/autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/ui/credential/management/AutofillManagementActivity.kt:114]: This is about to get worked on, will leave as a placeholder which will be imminently used

File: modules/autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/AutofillJavascriptInterface.kt [hunk 2]
```diff
@@ -46,10 +49,14 @@ class AutofillJavascriptInterface(
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply modules/autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/AutofillJavascriptInterface.kt:49]: Good question... I've added a cleanup method to null out the `WebView` when destroying the fragment

**Comment by CDRussell** [comment]
> 

I noticed this before too, and started an internal conversation about it

For now, I've removed a guard that re-enables the correct functionality. This might not be the _best_ fix, but it unblocks us for now. Can discuss the _better_ options, if any, in this task: https://app.asana.com/0/0/1202480731974698/f

**Pull request merged**